src/storage/src/frontend/mod.rs

```diff
@@ -155,21 +155,61 @@ impl<P: BackendStorage> FrontendStorage<P> {
         schema_name: &str,
         table_name: &str,
         rows: Vec<Vec<String>>,
+        columns: Option<Vec<String>>,
     ) -> SystemResult<Result<(), OperationOnTableError>> {
         match self.table_columns(schema_name, table_name)? {
             Ok(all_columns) => {
+                let index_columns = match columns {
+                    Some(cols) => {
+                        let mut index_cols = vec![];
+                        let mut non_existing_cols = vec![];
+                        for col in cols {
+                            let mut found = None;
+                            for (index, (name, sql_type)) in all_columns.iter().enumerate() {
+                                if *name == col {
+                                    found = Some((index, name.clone(), *sql_type));
+                                    break;
+                                }
+                            }
+
+                            match found {
+                                Some(index_col) => {
+                                    index_cols.push(index_col);
+                                }
+                                None => non_existing_cols.push(col),
+                            }
+                        }
+
+                        if !non_existing_cols.is_empty() {
+                            return Ok(Err(OperationOnTableError::ColumnDoesNotExist(non_existing_cols)));
+                        }
+
+                        index_cols
+                    }
+                    None => {
+                        let mut index_cols = vec![];
+                        for (index, (name, sql_type)) in all_columns.iter().enumerate() {
+                            index_cols.push((index, name.clone(), *sql_type));
+                        }
+
+                        index_cols
+                    }
+                };
+
                 let mut to_write: Vec<Row> = vec![];
                 let mut errors = HashMap::new();
                 for row in rows {
                     let key = self.key_id_generator.to_be_bytes().to_vec();
-                    let mut record = vec![];
+
+                    // TODO: The default value or NULL should be initialized for SQL types of all columns.
+                    let mut record = vec![vec![0, 0]; all_columns.len()];
```

Comment:
Why `vec![0, 0]` and not `vec![]` or `vec![0]`?

Reply:
I am sorry that I have not looked into the details. I tried to use `vec![]` and `vec![0]`, the `insert` operation succeeded, but the next `select` crashed. Below is the steps and crash log, I will look into it and update the PR.

```
xxx=> create schema schema_name;
CREATE SCHEMA
xxx=> create table schema_name.table_name (col1 smallint, col2 smallint, col3 smallint);
CREATE TABLE
xxx=> insert into schema_name.table_name (col3, col2) values (1, 2);
INSERT 0 1
asongala=> select * from schema_name.table_name;
```

For `vec![]`:

```
LEN = 21
PARAMS = b"client_encoding\0UTF8\0"
LEN = 26
PARAMS = b"S\0\0\0\x19client_encoding\0UTF8\0"
LEN = 14
PARAMS = b"DateStyle\0DMY\0"
LEN = 19
PARAMS = b"S\0\0\0\x12DateStyle\0DMY\0"
thread 'main' panicked at 'index 2 out of range for slice of length 0', src/sql_types/src/lib.rs:102:28
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

For `vec![0]`:

```
LEN = 21
PARAMS = b"client_encoding\0UTF8\0"
LEN = 26
PARAMS = b"S\0\0\0\x19client_encoding\0UTF8\0"
LEN = 14
PARAMS = b"DateStyle\0DMY\0"
LEN = 19
PARAMS = b"S\0\0\0\x12DateStyle\0DMY\0"
thread 'main' panicked at 'index 2 out of range for slice of length 1', src/sql_types/src/lib.rs:102:28
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Reply:
Thanks for posting errors. I got why it fails. This is because it assumes non null values. I hardcoded the size of possible data in storage here
https://github.com/alex-dukhno/database/blob/master/src/sql_types/src/lib.rs#L102.

I think we handle it when introducing `NOT NULL` constraint in #25 or #26

Reply:
Hi @alex-dukhno , should I try to fix it in another PR?